### PR TITLE
Verify proof of work for relayed messages

### PR DESCRIPTION
**Pull request by sachaaaaa**
Add `verify_message`, which checks ttl, timestamp, PoW and hash, in `process_push` and `process_push_batch`.
For batch, we remove the messages that are invalid but keep the ones valid.

### CODE REVIEW

File: utils/src/utils.cpp [hunk 1]
```diff
@@ -81,7 +89,7 @@ bool parseTTL(const std::string& ttlString, uint64_t& ttl) {
     }
 
     // Minimum time to live of 10 seconds, maximum of 4 days
```
Comment (msgmaxim) [inline utils/src/utils.cpp]: This comment belongs in validateTTL

File: httpserver/service_node.cpp [hunk 2]
```diff
@@ -93,6 +95,33 @@ static std::shared_ptr<request_t> make_push_request(std::string&& data) {
     return make_post_request("/v1/swarms/push", std::move(data));
 }
 
+template <class T>
+static bool verify_message(const T& msg, const char* error_message = nullptr) {
```
Comment (msgmaxim) [inline httpserver/service_node.cpp]: Why use a template when T is always `message_t`?

File: httpserver/service_node.cpp [hunk 3]
```diff
@@ -93,6 +95,33 @@ static std::shared_ptr<request_t> make_push_request(std::string&& data) {
...
+    if (!util::validateTTL(msg.ttl)) {
+        if (error_message)
+            error_message = "Provided TTL is not valid";
```
Comment (msgmaxim) [inline httpserver/service_node.cpp]: Not sure it does what you intended. You are modifying `error_message` here, not the original `error_msg`, correct? How about we just print a warning?

File: httpserver/service_node.cpp [hunk 4]
```diff
@@ -229,7 +258,14 @@ bool ServiceNode::process_store(const message_t& msg) {
     return true;
 }
 
-void ServiceNode::process_push(const message_t& msg) { save_if_new(msg); }
+void ServiceNode::process_push(const message_t& msg) {
+#ifndef DISABLE_POW
+    char* error_msg;
+    if (!verify_message(msg, error_msg))
+        throw std::runtime_error(error_msg);
```
Comment (msgmaxim) [inline httpserver/service_node.cpp]: Seems a bit too harsh. Just ignore the message, provided that we printed appropriate warnings?

File: httpserver/service_node.cpp (hunk 3 above)
Comment (sachaaaaa) [reply httpserver/service_node.cpp]: OOps wanted to pass the pointer by reference.
The idea is to provide a useful error message to the user, so the exception will bubble up until it's caught, and the exception message used in the response body.

File: httpserver/service_node.cpp (hunk 4 above)
Comment (sachaaaaa) [reply httpserver/service_node.cpp]: As said above, this way the error message bubbles up and is returned in the response body.

File: httpserver/service_node.cpp (hunk 2 above)
Comment (sachaaaaa) [reply httpserver/service_node.cpp]: True, didn't bother checking which struct was used here. Hopefully we soon only handle one struct for messages!

File: utils/src/utils.cpp (hunk 1 above)
Comment (sachaaaaa) [reply utils/src/utils.cpp]: Will fix.

File: httpserver/service_node.cpp (hunk 3 above)
Comment (sachaaaaa) [reply httpserver/service_node.cpp]: (obviously can't be a reference since I want it to be optional)

File: httpserver/service_node.cpp (hunk 4 above)
Comment (msgmaxim) [reply httpserver/service_node.cpp]: OK

**Pull request merged**